siphon/ncss.py

```diff
@@ -0,0 +1,391 @@
+def parse_xml(data, handle_units):
+    import xml.etree.ElementTree as ET
+    root = ET.fromstring(data)
+    return squish(parse_xml_dataset(root, handle_units))
+
+
+def parse_xml_point(elem):
+    point = dict()
+    units = dict()
+    for data in elem.findall('data'):
+        name = data.get('name')
+        unit = data.get('units')
+        point[name] = float(data.text) if name != 'date' else parse_iso_date(data.text)
+        if unit:
+            units[name] = unit
+    return point, units
+
+
+def combine_xml_points(l, units, handle_units):
+    ret = dict()
+    for item in l:
+        for key, value in item.items():
+            ret.setdefault(key, []).append(value)
+
+    for key, value in ret.items():
+        if key != 'date':
+            ret[key] = handle_units(value, units.get(key, None))
+
+    return ret
+
+
+def parse_xml_dataset(elem, handle_units):
+    points, units = zip(*[parse_xml_point(p) for p in elem.findall('point')])
+    # Group points by the contents of each point
+    datasets = dict()
+    for p in points:
+        datasets.setdefault(tuple(p.keys()), []).append(p)
+
+    all_units = combine_dicts(units)
+    return [combine_xml_points(d, all_units, handle_units) for d in datasets.values()]
+
+
+# Handling of netCDF 3/4 from NCSS
+try:
+    from netCDF4 import Dataset
+    from tempfile import NamedTemporaryFile
+
+    @response_handlers.register('application/x-netcdf')
+    @response_handlers.register('application/x-netcdf4')
+    def read_netcdf(data, handle_units):  # pylint:disable=unused-argument
+        with NamedTemporaryFile() as tmp_file:
+            tmp_file.write(data)
+            tmp_file.flush()
+            return Dataset(tmp_file.name, 'r')
+except ImportError:
+    import warnings
+    warnings.warn('netCDF4 module not installed. '
+                  'Will be unable to handle NetCDF returns from NCSS.')
+
+
+# Parsing of CSV data returned from NCSS
+@response_handlers.register('text/plain')
+def parse_csv_response(data, unit_handler):
+    return squish([parse_csv_dataset(d, unit_handler) for d in data.split(b'\n\n')])
+
+
+def parse_csv_header(line):
+    units = dict()
+    names = []
+    for var in line.split(','):
+        start = var.find('[')
+        if start < 0:
+            names.append(str(var))
+            continue
+        else:
+            names.append(str(var[:start]))
+        end = var.find(']', start)
+        unitstr = var[start + 1:end]
+        eq = unitstr.find('=')
+        if eq >= 0:
+            # go past = and ", skip final "
+            units[names[-1]] = unitstr[eq + 2:-1]
+    return names, units
+
+
+def parse_csv_dataset(data, handle_units):
+    fobj = BytesIO(data)
+    names, units = parse_csv_header(fobj.readline().decode('utf-8'))
+    arrs = np.genfromtxt(fobj, dtype=None, names=names, delimiter=',', unpack=True,
+                         converters={'date': lambda s: parse_iso_date(s.decode('utf-8'))})
```

Comment:
Have you ran into the funny (i.e. NOT funny) case of a csv return from a PointFeatureCollection that is missing the "T" character in the iso time string? While that should be fixed at some point [https://bugtracking.unidata.ucar.edu/browse/TDS-670], we may want to enhance parse_iso_date to handle that.